### PR TITLE
Implement matrix testing Python 3.10-3.12, ubuntu, windows, macOS(arm64 and intel)

### LINKED ISSUE #48: Implement matrix for _matrix-and-codecov-on-merge-to-main.yml (high priority)
We wanted to include matrix tests for ubuntu-latest, macos-latest, windows-latest, and macOS ARM64, if available. @Tieqiong

At the moment, we are testing this via conda-forge (not productive since we've already commited lots of code. Ideally, we need to know the exact compilers used in the build process so that we can emulate them in the conda-forge recipe.

### PR DESCRIPTION
**Pull request by bobleesj**
Closes #48 

Tested using a fork on diffpy.pdffit2: https://github.com/bobleesj/diffpy.pdffit2/actions/runs/11079705068/job/30789162458?pr=6

Screenshot 2024-09-27 at 9 06 34 PM

`macos-14` referes to macOS M1

@sbillinge ready for review

### CODE REVIEW

**Review by sbillinge — COMMENTED** [review]
Is there a reason we are pinning Ubuntu latest to 3.12? This is ok but it will just be more thinks to forget to do in the future when this shifts. I'm ok with leaving it but let's just give it some thought first

File: .github/workflows/_matrix-and-codecov-on-merge-to-main.yml [hunk 1]
```diff
@@ -29,7 +29,13 @@ jobs:
...
         shell: bash -l {0}
 
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-13, macos-14]
+        python-version: ["3.10", "3.11", "3.12"]
+    env:
+          LATEST_PYTHON_VERSION: "3.12"
```
Comment (bobleesj) [inline .github/workflows/_matrix-and-codecov-on-merge-to-main.yml:38]: What if we create `LATEST_PYTHON_VERSION` at the top? I did a quick search and found no simple way to extract the largest value from the `python-version` array.

**Review by bobleesj — COMMENTED** [review]
@sbillinge How do you like the new change to minimize future technical burden?

**Comment by Tieqiong** [comment]
@bobleesj there are several things we might want to change here. 
1: headless (xvfb) should only run for linux
2: arm Mac need to install functional packages, and often it's not the latest conda forage main release. (this should go away with a nice conda forge release)

I had a matrix workflow for pdfgui a while ago where I noticed the problems (also see Billingegroup/cookiecutter#109): https://github.com/diffpy/diffpy.pdfgui/actions/runs/10604930935/workflow

**Comment by bobleesj** [comment]
Thanks Tieqiong, it seems like an edge case for pdfgui out of all the packages we have.

for release scripts, I think it should be as general as possible, and for pdfgui, we can probably make a separate ci file for matrix test then? we won’t be distributing macOS arm64 (no native built) or Linux (wxpython) on conda forge

**Comment by Tieqiong** [comment]
Not only for pdfgui. Whenever we need a headless test we want only linux to run xvfb. So let's add that logic in, or we're removing headless. 

Also for the pdffit2 installation. I'm not sure but there could be other packages that have it as dependencies?

**Comment by bobleesj** [comment]
for linux headless, for sure, if you could make a PR for that. Would you also like to make a separate CI for pdfgui ? I don't recall any packages relying on pdffit2 besides pdfgui. @sbillinge maybe can provide insight on this

**Comment by sbillinge** [comment]
Per convo with Bob, let's wait until we are building wheel files and deplying on pypi and then do it that way.
